x/outline-cli/main.go

```diff
@@ -0,0 +1,390 @@
+// Copyright 2023 Jigsaw Operations LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"os/signal"
+	"strconv"
+	"time"
+
+	"github.com/Jigsaw-Code/outline-internal-sdk/network"
+	"github.com/Jigsaw-Code/outline-internal-sdk/network/dnstruncate"
+	"github.com/Jigsaw-Code/outline-internal-sdk/network/lwip2transport"
+	"github.com/Jigsaw-Code/outline-internal-sdk/transport"
+	"github.com/Jigsaw-Code/outline-internal-sdk/transport/shadowsocks"
+	"github.com/songgao/water"
+	"github.com/vishvananda/netlink"
+	"golang.org/x/sys/unix"
+)
+
+const OUTLINE_TUN_NAME = "outline233"
+const OUTLINE_TUN_IP = "10.233.233.1"
+const OUTLINE_TUN_MTU = 1500 // todo: we can read this from netlink
+const OUTLINE_TUN_SUBNET = "10.233.233.1/32"
+const OUTLINE_GW_SUBNET = "10.233.233.2/32"
+const OUTLINE_GW_IP = "10.233.233.2"
+const OUTLINE_ROUTING_PRIORITY = 23333
+const OUTLINE_ROUTING_TABLE = 233
+
+// ./app
+//
+//	<svr-ip>     : the outline server IP (e.g. 111.111.111.111)
+//	<svt-port>   : the outline server port (e.g. 21532)
+//	<svr-pass>   : the outline server password
+func main() {
+	fmt.Println("OutlineVPN CLI (experimental-07211507)")
+
+	svrIp := os.Args[1]
+	svrIpCidr := svrIp + "/32"
+	svrPass := os.Args[3]
+	svrPort, err := strconv.Atoi(os.Args[2])
+	if err != nil {
+		fmt.Printf("fatal error: %v\n", err)
+		return
+	}
+	if svrPort < 1000 || svrPort > 65535 {
+		fmt.Printf("fatal error: server port out of range\n")
+		return
+	}
+
+	// wait for go routine to terminate
+	defer time.Sleep(1 * time.Second)
+
+	tun, err := setupTunDevice()
+	if err != nil {
+		return
+	}
+	defer cleanUpTunDevice(tun)
+
+	if err := showTunDevice(); err != nil {
+		return
+	}
+	if err := configureTunDevice(); err != nil {
+		return
+	}
+	if err := showTunDevice(); err != nil {
+		return
+	}
+
+	err = setupRouting()
```

Comment:
Routing can be last.

I think the order is something like:
1. Create remote IPDevice (VPN), given a transport config. This should check connectivity.
2. Create local IPDevice (tun)
3. Bridge devices
4. Set up routing to use the local IPDevice (tun), given a routing config.

I'd like to see functions that makes that structure clearer.

I guess we need to add the device refresh somewhere there too